lib/sendgrid/mail.rb

```diff
@@ -28,34 +33,46 @@ def add_attachment(path, name = nil)
     end
 
     def to_h
-      payload = {
-        :from        => @from,
-        :fromname    => @from_name,
-        :subject     => @subject,
-        :to          => @to,
-        :toname      => @to_name,
-        :date        => @date,
-        :replyto     => @reply_to,
-        :cc          => @cc,
-        :bcc         => @bcc,
-        :text        => @text,
-        :html        => @html,
+      payload = {}
+
+      PAYLOAD_PARAMS.each do |payload_param|
+        payload[payload_mapping(payload_param)] = self.send(payload_param)
+      end
+
+      payload.merge!({
         :'x-smtpapi' => smtpapi_json,
         :files       => ({} unless @attachments.empty?)
-      }.reject {|k,v| v.nil?}
+      })
+
+      payload.reject! {|k,v| v.nil?}
+
+      assign_missing_to(payload)
+      extract_attachments(payload)
 
+      payload
+    end
+
+    def payload_mapping(key)
+      {
+        from_name: :fromname,
+        to_name: :toname,
+        reply_to: :replyto,
+      }[key] || key
+    end
+
+    def assign_missing_to(payload)
       # smtpapi fixer
       if @to.nil? and not @smtpapi.to.empty?
         payload[:to] = payload[:from]
```

Comment:
Is this correct? I'm not certain on the semantic value of these variables, but the names seem like this line should read `payload[:to] = @smtpapi.to`. What is this doing and when is it called?

Could you add the answer to that in a comment?